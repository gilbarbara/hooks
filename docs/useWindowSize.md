# useWindowSize

Get the window dimensions. Updates on resize.  

## Usage

```tsx
import { useWindowSize } from '@gilbarbara/hooks';

function Component() {
  const { width } = useWindowSize();

  return <div>{width >= 1024 ? 'Large Screen' : 'Small Screen'}</div>;
}
```

## Reference

```typescript
interface UseWindowSizeResult {
  height: number;
  width: number;
}

useWindowSize(debounce?: number): UseWindowSizeResult;
```
