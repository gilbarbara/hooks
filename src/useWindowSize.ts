import { useEffect, useRef, useState } from 'react';

import { canUseDOM, off, on } from './utils';

export interface UseWindowSizeResult {
  height: number;
  width: number;
}

export function useWindowSize(debounce = 0) {
  const [size, setSize] = useState<UseWindowSizeResult>({
    height: canUseDOM() ? window.innerHeight : 0,
    width: canUseDOM() ? window.innerWidth : 0,
  });
  const timeoutRef = useRef<number>(0);

  const handleResize = useRef(() => {
    window.clearTimeout(timeoutRef.current);

    timeoutRef.current = window.setTimeout(() => {
      setSize({
        width: window.innerWidth,
        height: window.innerHeight,
      });
    }, debounce);
  });

  useEffect(() => {
    /* c8 ignore next 3 */
    if (!canUseDOM()) {
      return () => undefined;
    }

    const getSize = handleResize.current;

    setSize({
      height: window.innerHeight,
      width: window.innerWidth,
    });
    on(window, 'resize', getSize);

    return () => {
      off(window, 'resize', getSize);
    };
  }, []);

  return size;
}
