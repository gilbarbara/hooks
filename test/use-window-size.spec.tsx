import * as React from 'react';
import { render } from '@testing-library/react';

import useWindowSize from '../src/use-window-size';

describe('useWindowSize', () => {
  const Component = () => {
    const size = useWindowSize();

    return <div>{JSON.stringify(size, null, 2)}</div>;
  };

  it('should show the changes', () => {
    const { container } = render(<Component />);

    expect(container.firstChild).toMatchSnapshot();
  });
});
