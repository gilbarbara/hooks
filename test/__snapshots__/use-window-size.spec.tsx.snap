// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`useWindowSize should show the changes 1`] = `
<div>
  {
  "width": 1024,
  "height": 768
}
</div>
`;
